File: grpc/hello-grpc-nodejs/proto_server.js
const grpc = require('@grpc/grpc-js')
let uuid = require('uuid')
let messages = require('./common/landing_pb')
let services = require('./common/landing_grpc_pb')
let conn = require('./common/connection')
//let ref = require('grpc-node-server-reflection')

const fs = require('fs')

let hellos = ["Hello", "Bonjour", "Hola", "こんにちは", "Ciao", "안녕하세요"]

let tracingKeys = [
    "x-request-id",
    "x-b3-traceid",
    "x-b3-spanid",
    "x-b3-parentspanid",
    "x-b3-sampled",
    "x-b3-flags",
    "x-ot-span-context"
]

const logger = conn.logger
let next
const cert = "/var/hello_grpc/server_certs/cert.pem"
const certKey = "/var/hello_grpc/server_certs/private.key"
const certChain = "/var/hello_grpc/server_certs/full_chain.pem"
const rootCert = "/var/hello_grpc/server_certs/myssl_root.cer"


/**
 * Starts an RPC server that receives requests for the Greeter service at the
 * sample server port
 */
function main() {
    if (hasBackend()) {
        next = conn.getClient()
    }

    let port = process.env.GRPC_SERVER_PORT
    if (typeof port == 'undefined' || port == null) {
        port = "9996"
    }
    let address = "0.0.0.0:" + port

    //const server =ref.default(new grpc.Server())
    let server = new grpc.Server()
    server.addService(services.LandingServiceService, {
        talk: talk,
        talkOneAnswerMore: talkOneAnswerMore,
        talkMoreAnswerOne: talkMoreAnswerOne,
        talkBidirectional: talkBidirectional
    })

    let secure = process.env.GRPC_HELLO_SECURE
    if (typeof secure !== 'undefined' && secure !== null && secure === "Y") {
        let checkClientCertificate = false;
        let rootCertContent = fs.readFileSync(rootCert);
        let certChainContent = fs.readFileSync(certChain);
        let privateKeyContent = fs.readFileSync(certKey);
        let credentials = grpc.ServerCredentials.createSsl(
            rootCertContent,
            [{cert_chain: certChainContent, private_key: privateKeyContent}],
            checkClientCertificate
        )
        server.bindAsync(address, credentials, () => {
            server.start()
            logger.info("Start GRPC TLS Server[%s]", port)
        })
    } else {
        server.bindAsync(address, grpc.ServerCredentials.createInsecure(), () => {
            server.start()
            logger.info("Start GRPC Server[%s]", port)
        })
    }
}

function talk(call, callback) {
    let request = call.request
    logger.info("TALK REQUEST: data=%s,meta=%s", request.getData(), request.getMeta())
    let metadata = propagandaHeaders("Talk", call)
    if (hasNext()) {
        next.talk(request, metadata, function (err, response) {
            callback(null, response)
        })
    } else {
        let response = new messages.TalkResponse()
        response.setStatus(200)
        const talkResult = buildResult(request.getData())
        let talkResults = [talkResult]
        response.setResultsList(talkResults)
        callback(null, response)
    }
}

function talkOneAnswerMore(call) {
    let request = call.request
    logger.info("TalkOneAnswerMore REQUEST: data=%s,meta=%s", request.getData(), request.getMeta())
    let metadata = propagandaHeaders("TalkOneAnswerMore", call)
    if (hasNext()) {
        let nextCall = next.talkOneAnswerMore(request, metadata)
        nextCall.on('data', function (response) {
            call.write(response)
        })
        nextCall.on('end', function () {
            call.end()
        })
    } else {
        let datas = request.getData().split(",")
        for (const data in datas) {
            let response = new messages.TalkResponse()
            response.setStatus(200)
            const talkResult = buildResult(data)
            let talkResults = [talkResult]
            response.setResultsList(talkResults)
            logger.info("TalkOneAnswerMore:%s", response)
            call.write(response)
        }
        call.end()
    }
}

function talkMoreAnswerOne(call, callback) {
    let metadata = propagandaHeaders("TalkMoreAnswerOne", call)
    if (hasNext()) {
        let nextCall = next.talkMoreAnswerOne(metadata, function (err, response) {
            if (err) {
                logger.error(err)
            } else {
                callback(null, response)
            }
        })
        call.on('data', function (request) {
            logger.info("TalkMoreAnswerOne REQUEST: data=%s,meta=%s", request.getData(), request.getMeta())
            nextCall.write(request)
        })
        call.on('end', function () {
            nextCall.end()
        })
        call.on('error', function (e) {
            logger.error(e)
        })
    } else {
        let talkResults = []
        call.on('data', function (request) {
            logger.info("TalkMoreAnswerOne REQUEST: data=%s,meta=%s", request.getData(), request.getMeta())
            talkResults.push(buildResult(request.getData()))
        })
        call.on('end', function () {
            let response = new messages.TalkResponse()
            response.setStatus(200)
            response.setResultsList(talkResults)
            callback(null, response)
        })
        call.on('error', function (e) {
            logger.error(e)
        })
    }
}

function talkBidirectional(call) {
    let metadata = propagandaHeaders("TalkBidirectional", call)
    if (hasNext()) {
        let nextCall = next.talkBidirectional(metadata)
        nextCall.on('data', function (response) {
            call.write(response)
        })
        nextCall.on('end', function () {
            call.end()
        })

        call.on('data', function (request) {
            nextCall.write(request)
        })
        call.on('end', function () {
            nextCall.end()
        })
    } else {
        call.on('data', function (request) {
            logger.info("TalkBidirectional REQUEST: data=%s,meta=%s", request.getData(), request.getMeta())
            let response = new messages.TalkResponse()
            response.setStatus(200)
            let data = request.getData()
            const talkResult = buildResult(data)
            let talkResults = [talkResult]
            response.setResultsList(talkResults)
            logger.info("TalkBidirectional:%s", response)
            call.write(response)
        })
        call.on('end', function () {
            call.end()
        })
    }
}

function hasBackend() {
    let backend = process.env.GRPC_HELLO_BACKEND
    return typeof backend !== 'undefined' && backend !== null
}

function hasNext() {
    return typeof next !== 'undefined' && next !== null;
}

// {"status":200,"results":[{"id":1600402320493411000,"kv":{"data":"Hello","id":"0"}}]}
function buildResult(id) {
    let result = new messages.TalkResult()
    let index = parseInt(id)
    result.setId(Math.round(Date.now() / 1000))
    result.setType(messages.ResultType.OK)
    let kv = result.getKvMap()
    kv.set("id", uuid.v1())
    kv.set("idx", id)
    kv.set("data", hellos[index])
    kv.set("meta", "NODEJS")
    return result
}

function propagandaHeaders(methodName, call) {
    let headers = call.metadata.getMap()
    const metadata = new grpc.Metadata()
    for (let key in headers) {
        logger.info("%s HEADER: %s:%s", methodName, key, headers[key])
        if (key in tracingKeys) {
            metadata.add(key, headers[key])
        }
    }
    return metadata
}

main()

File: grpc/hello-grpc-python/server/protoServer.py
# encoding: utf-8
import logging
import os
import sys
import time
import uuid

import grpc
from concurrent import futures

from conn import connection
from landing import landing_pb2
from landing import landing_pb2_grpc

logger = logging.getLogger('grpc-server')
logger.setLevel(logging.INFO)
console = logging.StreamHandler()
console.setLevel(logging.INFO)
formatter = logging.Formatter('%(asctime)s [%(levelname)s] - %(message)s')
console.setFormatter(formatter)
logger.addHandler(console)

hellos = ["Hello", "Bonjour", "Hola", "こんにちは", "Ciao", "안녕하세요"]
tracing_keys = [
    "x-request-id",
    "x-b3-traceid",
    "x-b3-spanid",
    "x-b3-parentspanid",
    "x-b3-sampled",
    "x-b3-flags",
    "x-ot-span-context"
]
cert = "/var/hello_grpc/server_certs/cert.pem"
certKey = "/var/hello_grpc/server_certs/private.key"
certChain = "/var/hello_grpc/server_certs/full_chain.pem"
rootCert = "/var/hello_grpc/server_certs/myssl_root.cer"


def build_result(data):
    result = landing_pb2.TalkResult()
    result.id = int((time.time()))
    result.type = landing_pb2.OK
    result.kv["id"] = str(uuid.uuid1())
    result.kv["idx"] = data
    index = int(data)
    result.kv["data"] = hellos[index]
    result.kv["meta"] = "PYTHON"
    return result


def propaganda_headers(method_name, context):
    metadata = context.invocation_metadata()
    metadata_dict = {}
    for c in metadata:
        logger.info("%s ->H %s:%s", method_name, c.key, c.value)
        if c.key in tracing_keys:
            logger.info("%s ->T %s:%s", method_name, c.key, c.value)
            metadata_dict[c.key] = c.value
    # Converting dictionary into list of tuple
    return list(metadata_dict.items())


def print_headers(method_name, context):
    metadata = context.invocation_metadata()
    for c in metadata:
        logger.info("%s ->H %s:%s", method_name, c.key, c.value)
        if c.key in tracing_keys:
            logger.info("%s ->T %s:%s", method_name, c.key, c.value)


class LandingServiceServer(landing_pb2_grpc.LandingServiceServicer):
    def __init__(self, next_one):
        self.next_one = next_one

    def Talk(self, request, context):
        logger.info("TALK REQUEST: data=%s,meta=%s", request.data, request.meta)
        if self.next_one:
            headers = propaganda_headers("TALK", context)
            try:
                return self.next_one.Talk(request=request, metadata=headers)
            except Exception as e:
                logger.error("Unexpected Error: {}".format(e))
        else:
            print_headers("TALK", context)
            response = landing_pb2.TalkResponse()
            response.status = 200
            result = build_result(request.data)
            response.results.append(result)
            return response

    def TalkOneAnswerMore(self, request, context):
        logger.info("TalkOneAnswerMore REQUEST: data=%s,meta=%s", request.data, request.meta)
        if self.next_one:
            headers = propaganda_headers("TalkOneAnswerMore", context)
            responses = self.next_one.TalkOneAnswerMore(request=request, metadata=headers)
            for response in responses:
                yield response
        else:
            print_headers("TalkOneAnswerMore", context)
            datas = request.data.split(",")
            for data in datas:
                response = landing_pb2.TalkResponse()
                response.status = 200
                result = build_result(data)
                response.results.append(result)
                yield response

    def TalkMoreAnswerOne(self, request_iterator, context):
        if self.next_one:
            headers = propaganda_headers("TalkMoreAnswerOne", context)
            return self.next_one.TalkMoreAnswerOne(request_iterator=request_iterator, metadata=headers)
        else:
            response = landing_pb2.TalkResponse()
            response.status = 200
            for request in request_iterator:
                logger.info("TalkMoreAnswerOne REQUEST: data=%s,meta=%s", request.data, request.meta)
                print_headers("TalkMoreAnswerOne", context)
                response.results.append(build_result(request.data))
            return response

    def TalkBidirectional(self, request_iterator, context):
        if self.next_one:
            headers = propaganda_headers("TalkBidirectional", context)
            responses = self.next_one.TalkBidirectional(request_iterator=request_iterator, metadata=headers)
            for response in responses:
                yield response
        else:
            for request in request_iterator:
                logger.info("TalkBidirectional REQUEST: data=%s,meta=%s", request.data, request.meta)
                print_headers("TalkMoreAnswerOne", context)
                response = landing_pb2.TalkResponse()
                response.status = 200
                result = build_result(request.data)
                response.results.append(result)
                yield response


def serve():
    backend = os.getenv("GRPC_HELLO_BACKEND")
    if backend:
        channel = connection.build_channel()
        stub = landing_pb2_grpc.LandingServiceStub(channel)
        # set next_one
        service_server = LandingServiceServer(stub)
    else:
        service_server = LandingServiceServer(None)
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    landing_pb2_grpc.add_LandingServiceServicer_to_server(service_server, server)

    port = os.getenv("GRPC_SERVER_PORT")
    if not port:
        port = "9996"
    address = "[::]:" + port
    secure = os.getenv("GRPC_HELLO_SECURE")
    python_version = sys.version_info
    if secure == "Y":
        # 以二进制格式打开一个文件用于只读
        with open(certKey, 'rb') as f:
            private_key = f.read()
        with open(certChain, 'rb') as f:
            certificate_chain = f.read()
        with open(rootCert, 'rb') as f:
            root_certificates = f.read()
        pairs = ((private_key, certificate_chain),)
        require_client_auth = False
        server_credentials = grpc.ssl_server_credentials(pairs, root_certificates, require_client_auth)
        server.add_secure_port(address, server_credentials)
        logger.info("Start GRPC TLS Server:[%s] (version:%s.%s.%s)", port, python_version[0], python_version[1],
                    python_version[2])
    else:
        server.add_insecure_port(address)
        logger.info("Start GRPC Server:[%s] (version:%s.%s.%s)", port, python_version[0], python_version[1],
                    python_version[2])
    server.start()

    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        server.stop(0)
    channel.close()


if __name__ == '__main__':
    serve()
